instance/models/load_balancer.py

```diff
@@ -111,6 +145,55 @@ def __init__(self, *args, **kwargs):
     def __str__(self):
         return self.domain
 
+    def modifies_configuration(method):  # nopep8 # pylint: disable=no-self-argument
+        """
+        Flag an internal method as one that modifies the state of load balancer configuration,
+        to allow managing configuration state outside and around the method call.
+        """
+        @functools.wraps(method)
+        def wrapper(self, *args, **kwargs):
+            """
+            Used to wrap the function that modified load balancer configuration state.
+            """
+            try:
+                # Prevent race conditions by updating state only if the LB is actually clean or dirty.
+                updated = LoadBalancingServer.objects.filter(
+                    pk=self.pk,
+                    configuration_state__in=[self.CLEAN, self.DIRTY]
+                ).update(
+                    configuration_state=self.IN_PROGRESS
+                )
+
+                # A race condition or just a normal queuing up of LB reconfiguration requests is handled.
+                if not updated:
+                    raise OtherReconfigurationInProgress
+
+                method(self, *args, **kwargs)  # pylint: disable=not-callable
+            except OtherReconfigurationInProgress:
+                # Prevent a race condition that gets the LB stuck in 'in-progress-and-dirty'
+                # if reconfiguration fails and configuration state is set to dirty, at the same
+                # time that a thread reaches here, by ensuring reconfiguration is really in progress.
+                LoadBalancingServer.objects.filter(
+                    pk=self.pk,
+                    configuration_state=self.IN_PROGRESS
+                ).update(
+                    configuration_state=self.IN_PROGRESS_AND_DIRTY
+                )
+            except ReconfigurationFailed:
+                # If the active reconfiguration failed for any reason, the LB is dirty and ready
+                # for another round of reconfiguration, either by request or periodic task.
+                self.configuration_state = self.DIRTY
+            else:
+                self.configuration_state = (
+                    self.DIRTY
+                    if self.configuration_state == self.IN_PROGRESS_AND_DIRTY
+                    else self.CLEAN
+                )
```

Comment:
`self.configuration_state` may be out of date here.  At the very least, we need to do `self.refresh_from_db()` before doing that check, but we still would have a race condition.  This needs to become an atomic update, like the one above, but I'm not sure that can be expressed in Django without using a database transaction (i.e. the `atomic` context manager).

Reply:
See above -- we can maybe just use a queryset.

Reply:
The problem here is the if condition.  You need two updates to express this in the same way as above:

    LoadBalancingServer.objects.filter(
        configuration_state=self.IN_PROGRESS
    ).update(
        configuration_state=self.CLEAN
    )
    LoadBalancingServer.objects.filter(
        configuration_state=self.IN_PROGRESS_AND_DIRTY
    ).update(
        configuration_state=self.DIRTY
    )

Because it's two updates, they are not atomic anymore.  If you put them in the order given above, it would actually work, but if you swap them, you would have a race condition again.

Anyway, we might not even stick with this solution, so we don't need to further discuss it.